### PR TITLE
Validates listen-gossip and listen-http options

### LINKED ISSUE #1805: listen options should provide hints if the user forgets the IP address
In an effort to be the humanely helpful software Habitat strives to be, we should be nicer when a user specifies the `--listen-http` and/or `--listen-gossip` CLI options to `hab start` but does not specify the address.

The required format of the command is:

```
root@localhost# hab start core/redis --listen-http 0.0.0.0:9700 --listen-gossip 0.0.0.0:7900
```

However, if the listen address is not specified (`0.0.0.0`), we get this message:

```
root@localhost# hab start core/redis --listen-http 9700 --listen-gossip 7900
hab-sup(CFG)[src/config.rs:74:34]: Failed to discover this hosts outbound IP address
```

We could do a couple things here:

1. Display a nice message explaining the problem and which option had an issue, e.g.:
```
hab-sup(CFG)[src/config.rs:##/##]: Failed to discover this host's outbound IP address, specify --listen-http IP:PORT
hab-sup(CFG)[src/config.rs:##/##]: Failed to discover this host's outbound IP address, specify --listen-gossip IP:PORT
```
2. Assume benign intent and listen on `0.0.0.0`. This is probably what the user meant if they don't specify an address

I think we should definitely do 1, and optionally do 2.

### PR DESCRIPTION
**Pull request by rsertelon**
Should close #1805

Signed-off-by: Romain Sertelon <romain@sertelon.fr>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/sup/src/main.rs [hunk 1]
```diff
@@ -1460,6 +1460,13 @@ fn valid_topology(val: String) -> result::Result<(), String> {
     }
 }
 
+fn valid_listen(val: String) -> result::Result<(), String> {
+    match GossipListenAddr::from_str(&val) {
+        Ok(_) => Ok(()),
+        Err(_) => Err(format!("Listen address should include both IP and port, eg: '0.0.0.0:9700'"))
+    }
+}
```
Comment (christophermaier) [inline components/sup/src/main.rs]: Though both things that you're validating are IP/port addresses, in the codebase they are represented by distinct types. The `LISTEN_GOSSIP` options are `GossipListenAddr`s, while the `LISTEN_HTTP` options are `http_gateway::ListenAddr`s.

For clarity, they should both have their own validators, which use the appropriate types (even though they are pretty much going to be the same functions).

Comment (rsertelon) [reply components/sup/src/main.rs]: Oh, yes, didn't know. Will do when I'll be on my computer. Thanks!

**Comment by rsertelon** [comment]
@christophermaier I've changed as asked :)

**Comment by christophermaier** [comment]
tenor-140505601

@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/296443078).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/296443078).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**